rfcs/2020-10-15-3480-file-source-rework.md

```diff
@@ -0,0 +1,425 @@
+exposes methods very much like the ones mentioned above. This would maintain the
+current structure of one main read loop with multiple points of control. An
+alternative would be to introduce a trait representing the logic of the read
+loop. This may lead to some duplication, but would allow simpler separation of
+different use cases and require readers of the code understand the interaction
+between fewer subtle points of control.
+
+I would propose that we begin with the simpler method of consolidating logic
+from the existing structure. This involves fewer design decisions and we may
+find with the other simplifications we're planning make that further simplifying
+the read loop is not worthwhile. That being said, once related features (like
+the batch mode discussed later) are complete, we can reevaluate this decision.
+
+#### File identity
+
+Identity is an area where we already have the beginning of modularity with the
+`Fingerprinter` abstraction, but it's incomplete. In our pseudocode, identity is
+involved in both `look_for_files` and `reconcile`. It's not just a way to figure
+out a magic identifier, but also the logic to update our list of watched files
+based on those identifiers. It needs to answer the following:
+
+1. Given a visible path, does it contain a file I've seen before?
+1. If I have seen it before, has it been renamed?
+1. If I have seen it before, am I now seeing it in multiple places?
+1. If I'm seeing duplicates, how should I choose which to follow?
+
+This logic is mostly in one place in the current implementation, so it should
+not be terribly difficult to extract it. The use of `Fingerprinter` should
+likely become purely internal to the new component.
+
+In addition to simply consolidating the logic, we can expand and make the use of
+`Fingerprinter` more intelligent. We currently have three ways it can work:
+
+1. Checksum (usually reliable but frustrating for small files)
+1. Device and inode (simple and works with small files, but doesn't handle edge
+   cases well)
+1. First line checksum (solid for intended use case but not yet general)
+
+I'd first propose that we drop device and inode fingerprinting and add
+path-based fingerprinting in its place. This gives users the option to do the
+simplest possible thing for use cases that don't need to worry about traditional
+rotation.
+
+Next, I suggest we unify the two checksumming strategies. Neither is perfectly
+general, but we should be able to come up with a simple algorithm that combines
+the best of both. As a prerequisite, we should unify the read path such that
+these checksums handle compressed files correctly (discussed later). With that
+in place, the algorithm can look something like the following:
+
+1. Read up to `max_line_length` bytes from the file starting at
+   `ignored_header_bytes`
+1. Return no fingerprint if there is no newline in the returned bytes
+1. Otherwise, return the checksum of the bytes up to the first newline
```

Comment:
Big fan of this. I'm curious what @MOZGIII's succint thoughts are for the Kubernetes use case? :)

Reply:
This is pretty much what the `FirstLineChecksum` does: https://github.com/timberio/vector/blob/c43b5e5928101cc45b3d5dea56211e56003e3465/lib/file-source/src/file_server.rs#L449-L450

So, yeah, this will just work!

Reply:
Yeah, the main differences between this and the existing `FirstLineChecksum` are basically:

1. Put an upper bound on the number of bytes read
2. Allow skipping over common headers
3. Work behind any compression handling

So the same idea, just with some tweaks to be more general.

Reply:
👍 

(1) is also already there, we have a limit on the amount of data before we give up trying to find the `\n` and just proceed with the checksum with what we've read so far.